File: CulinarySiteApp/src/app/services/author.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { IAuthorDetailModel } from '../interfaces/author/author-detail-model.interface';
import { CreateAuthorModel } from '../viewmodels/author/create-author-model.class';
import { UpdateAuthorModel } from '../viewmodels/author/update-author-model.class';
import { IAuthorListModel } from '../interfaces/author/author-list-model.interface';

@Injectable()
export class AuthorService {
  public url: string = '/api/author';

  constructor(private http: HttpClient) {}

  public getAuthorList(): Observable<IAuthorListModel[]> {
    return this.http.get<IAuthorListModel[]>(this.url);
  }

  public getAuthor(
    id: number,
    withRelated: boolean
  ): Observable<IAuthorDetailModel> {
    return this.http.get<IAuthorDetailModel>(
      this.url + '/' + id + '/' + withRelated
    );
  }

  public createAuthor(createAuthorModel: CreateAuthorModel): Observable<void> {
    return this.http.post<void>(this.url, createAuthorModel);
  }

  public updateAuthor(updateAuthorModel: UpdateAuthorModel): Observable<void> {
    return this.http.put<void>(this.url, updateAuthorModel);
  }

  public deleteAuthor(id: number): Observable<void> {
    return this.http.delete<void>(this.url + '/' + id);
  }
}

File: CulinarySite/CulinarySite.Bll/Services/AuthorService.cs
﻿using System.Collections.Generic;
using AutoMapper;
using CulinarySite.Bll.Interfaces;
using CulinarySite.Common.Dtos.Author;
using CulinarySite.Dal.Interfaces;
using CulinarySite.Domain.Entities;
using System.Linq;
using CulinarySite.Common.Exceptions;

namespace CulinarySite.Bll.Services
{
    public class AuthorService : IAuthorService
    {
        private readonly IReadOnlyGenericRepository<Author> _authorReadOnlyRepository;
        private readonly IWriteGenericRepository<Author> _authorWriteRepository;
        private readonly IMapper _mapper;
        public AuthorService(
            IReadOnlyGenericRepository<Author> authorReadOnlyRepository,
            IWriteGenericRepository<Author> authorWriteRepository,
            IMapper mapper)
        {
            _authorReadOnlyRepository = authorReadOnlyRepository;
            _authorWriteRepository = authorWriteRepository;
            _mapper = mapper;
        }

        public void CreateAuthor(CreateAuthorDto createAuthorDto)
        {
            var authorNames = _authorReadOnlyRepository.GetItemList().Select(x => x.Name);

            if (authorNames.Contains(createAuthorDto.Name))
            {
                throw new ValidationException($"The author with name:{createAuthorDto.Name} already exists.");
            }

            Author author = _mapper.Map<Author>(createAuthorDto);

            _authorWriteRepository.Create(author);
            _authorWriteRepository.Save();
        }

        public void UpdateAuthor(UpdateAuthorDto updateAuthorDto)
        {           
            Author author = _mapper.Map<Author>(updateAuthorDto);

            _authorWriteRepository.Update(author);
            _authorWriteRepository.Save();
        }

        public void DeleteAuthor(int id)
        {
            _authorWriteRepository.Delete(id);
            _authorWriteRepository.Save();
        }

        public AuthorDetailDto GetAuthor(int id, bool withRelated)
        {
            var author = new Author();
            var authorDetailDto = new AuthorDetailDto();

            if (withRelated)
            {
                author = _authorReadOnlyRepository.GetItemWithInclude(
                       x => x.Id == id,
                       x => x.Books,
                       x => x.Recipes);

                authorDetailDto = _mapper.Map<AuthorDetailDto>(author);

                return authorDetailDto;
            }

            author = _authorReadOnlyRepository.GetItem(id);

            authorDetailDto = _mapper.Map<AuthorDetailDto>(author);

            return authorDetailDto;
        }

        public IEnumerable<AuthorListDto> GetAuthorList()
        {
            IEnumerable<Author> authors = _authorReadOnlyRepository.GetItemList();
            var authorListDtos = authors.Select(x => _mapper.Map<AuthorListDto>(x));
           
            return authorListDtos;
        }

    }
}
